packages/std/src/math/decimal.rs

```diff
@@ -55,6 +56,30 @@ impl Decimal {
     pub fn is_zero(&self) -> bool {
         self.0 == 0
     }
+
+    /// Returns the approximate square root as a Decimal.
+    ///
+    /// This should not overflow or panic.
+    pub fn sqrt(&self) -> Self {
+        // We start with the highest precision possible and lower it until
+        // there's no overflow.
+        (0..=9)
+            .rev()
+            .find_map(|i| self.sqrt_with_precision(i))
+            .unwrap()
+    }
+
+    /// Lower precision means more aggressive rounding, but less risk of overflow.
+    /// Precision *must* be a number between 0 and 9 (inclusive).
+    ///
+    /// Returns `None` if the internal multiplication overflows.
+    fn sqrt_with_precision(&self, precision: u32) -> Option<Self> {
+        let inner_mul = 100u128.pow(precision);
+        let outer_mul = 10u128.pow(9 - precision);
```

Comment:
Is Rust clever enough to move the `outer_mul` calculation into the closure and more execute when `checked_mul` fails?

Reply:
I have no idea to be honest. Let's change this up.